Sporadic OkHttp errors after upgrading to ktor 1.3.1

**Ktor Version and Engine Used (client or server and name)**

```
implementation("io.ktor:ktor-client-core-jvm:1.3.1")
implementation("io.ktor:ktor-client-core:1.3.1")
implementation("io.ktor:ktor-client-jackson:1.3.1")
implementation("io.ktor:ktor-client-logging-jvm:1.3.1")
implementation("io.ktor:ktor-client-okhttp:1.3.1")
implementation("io.ktor:ktor-jackson:1.3.1")
implementation("io.ktor:ktor-metrics-micrometer:1.3.1")
implementation("io.ktor:ktor-metrics:1.3.1")
implementation("io.ktor:ktor-server-host-common:1.3.1")
implementation("io.ktor:ktor-server-netty:1.3.1")
```

**Describe the bug**

We've started seeing sporadic OkHttp exceptions in our tests when we upgraded from `1.3.0` to `1.3.1`. Upon downgrading back to `1.3.0` (with no other changes) the tests were fine again.

The exception is:

```
java.io.EOFException: \n not found: limit=0 content=…
	at o.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:240)
	at o.i.h.Http1ExchangeCodec.readHeaderLine(Http1ExchangeCodec.java:242)
	at o.i.h.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:213)
	... 20 common frames omitted
Wrapped by: java.io.IOException: unexpected end of stream on http://called-service/...
	at o.i.h.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:236)
	at o.i.c.Exchange.readResponseHeaders(Exchange.java:115)
	at o.i.h.CallServerInterceptor.intercept(CallServerInterceptor.java:94)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at o.i.c.ConnectInterceptor.intercept(ConnectInterceptor.java:43)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at o.i.c.CacheInterceptor.intercept(CacheInterceptor.java:94)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at o.i.h.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at o.i.h.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at o.i.h.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:221)
	at o.RealCall$AsyncCall.execute(RealCall.java:172)
	at o.i.NamedRunnable.run(NamedRunnable.java:32)
	at j.u.c.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at j.u.c.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:834)
```

It looks like the HTTP client receives an EOF while reading the response headers from calling http://called-service/... I don't think the culprit is the remote web server though, because this server is called from other services and it's only the one with ktor 1.3.1 that throws these errors. And this one is also fine with ktor 1.3.0.

This issue has already been raised before in the OkHttp repository, for example [here](https://github.com/square/okhttp/issues/5390).

From what I can see, both ktor `1.3.0` and `1.3.1` both depend on `com.squareup.okhttp3:okhttp:3.14.2`, so maybe this is due to a change in the way ktor uses OkHttp?

**To Reproduce**

Not easy to reproduce, unfortunately, because the error only occurs in a small percentage of requests (~5%). I'm hoping someone else observes this issue and can come up with a reproducible test.

I just got this error on [this app](https://github.com/brewin/mvi-coroutines) with 1.3.2 using OkHttp. It's seemingly random. Maybe once every 50 requests.

```
java.io.IOException: unexpected end of stream on https://api.github.com/...
        at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:236)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:221)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.java:172)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)
     Caused by: java.io.EOFException: \n not found: limit=0 content=…
        at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:240)
        at okhttp3.internal.http1.Http1ExchangeCodec.readHeaderLine(Http1ExchangeCodec.java:242)
        at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:213)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115) 
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) 
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) 
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) 
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) 
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) 
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) 
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:221) 
        at okhttp3.RealCall$AsyncCall.execute(RealCall.java:172) 
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:923) 
```

--- a/ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpConfig.kt
+++ b/ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpConfig.kt
@@ -15,7 +15,8 @@ public class OkHttpConfig : HttpClientEngineConfig() {
     internal var config: OkHttpClient.Builder.() -> Unit = {
         followRedirects(false)
         followSslRedirects(false)
-        retryOnConnectionFailure(false)
+
+        retryOnConnectionFailure(true)
     }
 
     /**

--- a/ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkUtils.kt
+++ b/ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkUtils.kt
@@ -68,23 +68,23 @@ internal fun Protocol.fromOkHttp(): HttpProtocolVersion = when (this) {
     Protocol.QUIC -> HttpProtocolVersion.QUIC
 }
 
-private fun mapOkHttpException(requestData: HttpRequestData, origin: IOException) =
-    when (val cause = origin.unwrapOkHttpCancelledException()) {
-        is SocketTimeoutException ->
-            if (cause.isConnectException()) {
-                ConnectTimeoutException(requestData, cause)
-            } else {
-                SocketTimeoutException(requestData, cause)
-            }
-        else -> cause
-    }
+private fun mapOkHttpException(
+    requestData: HttpRequestData,
+    origin: IOException
+): Throwable = when (val cause = origin.unwrapSuppressed()) {
+    is SocketTimeoutException ->
+        if (cause.isConnectException()) {
+            ConnectTimeoutException(requestData, cause)
+        } else {
+            SocketTimeoutException(requestData, cause)
+        }
+    else -> cause
+}
 
 private fun IOException.isConnectException() =
     message?.contains("connect", ignoreCase = true) == true
 
-private fun IOException.unwrapOkHttpCancelledException() =
-    if (message?.contains("canceled due to ") == true && suppressed.isNotEmpty()) {
-        suppressed[0]
-    } else {
-        this
-    }
+private fun IOException.unwrapSuppressed(): Throwable {
+    if (suppressed.isNotEmpty()) return suppressed[0]
+    return this
+}

